### PR TITLE
MB-22: Sign ether transaction

### PR DESCRIPTION
**Pull request by impul**
### Jira references
- Ticket (issue) — https://essentiatechnologies.atlassian.net/browse/MB-22
### Description
Add ability to sign `Ethereum` transaction to `Wallet`. 
### Implemented
Add raw transaction model:
```Swift
public struct RawTransaction: Codable {
    /// Amount value to send, unit is in Wei
    public let value: Wei
    /// Address to send ether to
    public let to: Address
    /// Gas price for this transaction, unit is in Wei
    /// you need to convert it if it is specified in GWei
    /// use Converter.toWei method to convert GWei value to Wei
    public let gasPrice: Int
    /// Gas limit for this transaction
    /// Total amount of gas will be (gas price * gas limit)
    public let gasLimit: Int
    /// Nonce of your address
    public let nonce: Int
    /// Data to attach to this transaction
    public let data: Data
}
```
Add method in sign raw transaction:
```Swift
public func sign(rawTransaction: RawTransaction) throws -> String 
```
### References
- Sign transaction — https://github.com/D-Technologies/EthereumKit/blob/master/Documentation/GettingStarted.md

### CODE REVIEW

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift [hunk 1]
```diff
@@ -24,21 +24,26 @@ public final class ECDSA {
         var pKey = secp256k1_pubkey()
         
         var result = secp256k1_ec_pubkey_create(context, &pKey, prvKey)
-        if (result != 1) {
+        if result != 1 {
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: @impul 1? add name for this magic number

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift [hunk 2]
```diff
@@ -24,21 +24,26 @@ public final class ECDSA {
...
+        let copressingKey = isCompression ? UInt32(SECP256K1_EC_COMPRESSED) : UInt32(SECP256K1_EC_UNCOMPRESSED)
         
-        result = secp256k1_ec_pubkey_serialize(context, pubkey, &s, &pKey, isCompression ? UInt32(SECP256K1_EC_COMPRESSED) : UInt32(SECP256K1_EC_UNCOMPRESSED))
-        if (result != 1) {
+        result = secp256k1_ec_pubkey_serialize(context,
+                                               pubkey,
+                                               &sizeT,
+                                               &pKey,
+                                               copressingKey)
+        if result != 1 {
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: @impul ,see comment above

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift [hunk 3]
```diff
@@ -24,21 +24,26 @@ public final class ECDSA {
         var pKey = secp256k1_pubkey()
         
         var result = secp256k1_ec_pubkey_create(context, &pKey, prvKey)
-        if (result != 1) {
+        if result != 1 {
             return nil
         }
         
         let size = isCompression ? 33 : 65
         let pubkey = UnsafeMutablePointer<UInt8>.allocate(capacity: size)
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: @impul , unclear with pubkey. there is one pKey before, which is public key. please specify name for these vars

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift [hunk 4]
```diff
@@ -24,21 +24,26 @@ public final class ECDSA {
         var pKey = secp256k1_pubkey()
         
         var result = secp256k1_ec_pubkey_create(context, &pKey, prvKey)
-        if (result != 1) {
+        if result != 1 {
             return nil
         }
         
         let size = isCompression ? 33 : 65
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: @impul 33, 65 -> think this is compressionBase

File: HDWalletKit/Core/Crypto/Encryption/EIP155Signer.swift [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import CryptoSwift
+
+public struct EIP155Signer {
+    
+    private let chainID: Int = 1
+    
+    public func sign(_ rawTransaction: RawTransaction, privateKey: PrivateKey) throws -> Data {
+        let transactionHash = try hash(rawTransaction: rawTransaction)
+        let signiture = try privateKey.sign(hash: transactionHash)
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Core/Crypto/Encryption/EIP155Signer.swift]: @impul signature. Pls check all this file

File: HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        return status == 1 ? signature : nil
+    }
+    
+    /// Converts signature data structure to 65 bytes.
+    ///
+    /// - Parameter signature: signature data structure
+    /// - Returns: 65 byte exported signature data.
+    public func export(signature: inout secp256k1_ecdsa_recoverable_signature) -> Data {
+        var output = Data(count: 65)
+        var recid = 0 as Int32
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift]: @impul what is recid?

File: HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    /// - Parameters:
+    ///   - signature: signature structure
+    ///   - hash: 32-byte (256-bit) hash of a message
+    /// - Returns: public key structure or nil, if signature invalid
+    public func publicKey(signature: inout secp256k1_ecdsa_recoverable_signature, hash: Data) -> secp256k1_pubkey? {
+        precondition(hash.count == 32, "Hash must be 32 bytes size")
+        let hash = hash.bytes
+        var outPubKey = secp256k1_pubkey()
+        let status = secp256k1_ecdsa_recover(context, &outPubKey, &signature, hash)
+        return status == 1 ? outPubKey : nil
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift]: @impul , magic number

File: HDWalletKit/Models/Address.swift [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    public let string: String
+    
+    public init(data: Data, string: String) {
+        self.data = data
+        self.string = string
+    }
+    
+    public init(data: Data) {
+        self.data = data
+        self.string = "0x" + EIP55.encode(data)
```
Comment (kozlov-oleksandr) [inline HDWalletKit/Models/Address.swift]: @impul ETH? BTC address has no `0x`

File: HDWalletKit_Tests/CryptoTests.swift [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+//
+//  Created by Pavlo Boiko on 12.07.18.
+//  Copyright © 2018 Essentia. All rights reserved.
+//
+
+import XCTest
+@testable import HDWalletKit
+
+class CryptoTests: XCTestCase {
+    func testSha3keccak256() {
```
Comment (kozlov-oleksandr) [inline HDWalletKit_Tests/CryptoTests.swift]: @impul testSHA3Keccak256

File: HDWalletKit_Tests/CryptoTests.swift [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    func testRestoringSignatureSignedWithOldScheme() {
+        let v = 27
+        let r = "75119860711638973245538703589762310947594328712729260330312782656531560398776"
+        let s = "51392727032514077370236468627319183981033698696331563950328005524752791633785"
+        let signer = EIP155Signer()
+        let signature = signer.calculateSignature(r: BInt(r)!, s: BInt(s)!, v: BInt(v))
+        XCTAssertEqual(signature.toHexString(), "a614559de76862bb1dbf8a969d8979e5bf21b72c51c96b27b3d247b728ebffb8719f40b018940ffd0880285d2196cdd31a710bf7cdda60c77632743d687dff7900")
+    }
+    
+    func testRestoringSignatureSignedWithOldScheme2() {
```
Comment (kozlov-oleksandr) [inline HDWalletKit_Tests/CryptoTests.swift]: @impul join this and previous tests in one

**Review by kozlov-oleksandr — CHANGES_REQUESTED** [review]

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift (hunk 1 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: done

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift (hunk 2 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: done

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift (hunk 3 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: done

File: HDWalletKit/Core/Crypto/Encryption/ECDSA.swift (hunk 4 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/ECDSA.swift]: done

File: HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift (hunk 6 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift]: @kozlov-oleksandr, recid - a pointer to an integer to hold the recovery id

File: HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift (hunk 7 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift]: done

File: HDWalletKit/Models/Address.swift (hunk 8 above)
Comment (impul) [reply HDWalletKit/Models/Address.swift]: done

File: HDWalletKit_Tests/CryptoTests.swift (hunk 9 above)
Comment (impul) [reply HDWalletKit_Tests/CryptoTests.swift]: done

File: HDWalletKit_Tests/CryptoTests.swift (hunk 10 above)
Comment (impul) [reply HDWalletKit_Tests/CryptoTests.swift]: done

File: HDWalletKit/Core/Crypto/Encryption/EIP155Signer.swift (hunk 5 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/EIP155Signer.swift]: done

File: HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift (hunk 6 above)
Comment (kozlov-oleksandr) [reply HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift]: @impul recId

**Review by kozlov-oleksandr — APPROVED** [review]

File: HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift (hunk 6 above)
Comment (impul) [reply HDWalletKit/Core/Crypto/Encryption/EllipticCurveEncrypterSecp256k1.swift]: @kozlov-oleksandr done

**Pull request merged**